smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,308 @@
+                    ModularXmppClientToServerConnectionInternal connectionInternal) {
+        return new EstablishingWebsocketConnectionState(establishingWebsocketConnectionStateDescriptor,
+                        connectionInternal);
+    }
+
+    public static final class WebsocketConnectedResult extends StateTransitionResult.Success {
+        final WebsocketRemoteConnectionEndpoint connectedEndpoint;
+
+        public WebsocketConnectedResult(WebsocketRemoteConnectionEndpoint connectedEndpoint) {
+            super("Websocket connection establised with endpoint: " + connectedEndpoint.getWebsocketEndpoint());
+            this.connectedEndpoint = connectedEndpoint;
+        }
+    }
+
+    private DiscoveredWebsocketEndpoints discoveredWebsocketEndpoints;
+
+    /**
+     * Transport class for {@link ModularXmppClientToServerConnectionModule}'s websocket implementation.
+     */
+    public final class XmppWebsocketTransport extends XmppClientToServerTransport {
+
+        protected XmppWebsocketTransport(ModularXmppClientToServerConnectionInternal connectionInternal) {
+            super(connectionInternal);
+        }
+
+        @Override
+        protected void resetDiscoveredConnectionEndpoints() {
+            discoveredWebsocketEndpoints = null;
+        }
+
+        @Override
+        protected List<SmackFuture<LookupConnectionEndpointsResult, Exception>> lookupConnectionEndpoints() {
+            // Assert that there are no stale discovered endpoints prior performing the lookup.
+            assert discoveredWebsocketEndpoints == null;
+
+            List<SmackFuture<LookupConnectionEndpointsResult, Exception>> futures = new ArrayList<>(2);
+
+            InternalSmackFuture<LookupConnectionEndpointsResult, Exception> websocketEndpointsLookupFuture = new InternalSmackFuture<>();
+
+            connectionInternal.asyncGo(() -> {
+
+                WebsocketRemoteConnectionEndpoint providedEndpoint = null;
+
+                // Check if there is a websocket endpoint already configured.
+                if (moduleDescriptor.getUri() != null) {
+                    providedEndpoint = new WebsocketRemoteConnectionEndpoint(moduleDescriptor.getUri());
+                }
+
+                if (!moduleDescriptor.isWebsocketEndpointDiscoveryEnabled()) {
+                    // If discovery is disabled, assert that the provided endpoint isn't null.
+                    assert providedEndpoint != null;
+
+                    if ((providedEndpoint.isSecureEndpoint() &&
+                            connectionInternal.connection.getConfiguration().getSecurityMode().equals(SecurityMode.disabled))
+                            || (!providedEndpoint.isSecureEndpoint() &&
+                                    connectionInternal.connection.getConfiguration().getSecurityMode().equals(SecurityMode.required))) {
+                        throw new IllegalStateException("Endpoint provided does not satisfy the security mode configuration");
```

Comment:
The exception message is not really user friendly, as it leaves the user uncertain what is actually not satisfied. :-(

Smack should return a clear message what is wrong, either that SecurityMode is disabled, but we only got a secure websocket URI, or vice versa.